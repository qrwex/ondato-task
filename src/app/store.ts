import {configureStore, combineReducers} from '@reduxjs/toolkit';
import {persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import clientsReducer from '../features/clients/clientsSlice';

const rootReducer = combineReducers({
  clients: clientsReducer,
});

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
