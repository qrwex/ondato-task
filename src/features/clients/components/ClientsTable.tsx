import {
  IconButton,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from "@material-ui/core";
import EditIcon from "@material-ui/icons/Edit";
import { selectAll } from "../clientsSlice";
import DeleteIcon from "@material-ui/icons/Delete";
import { useAppSelector } from "../../../app/hooks";
import { FC } from "react";

interface ClientsTableProps {
  onDelete: (id: string) => void;
  onModify: (id: string) => void;
}

export const ClientsTable: FC<ClientsTableProps> = ({ onDelete, onModify }) => {
  const all = useAppSelector(selectAll);

  return (
    <Table>
      <TableHead>
        <TableRow>
          <TableCell>#</TableCell>
          <TableCell>First name</TableCell>
          <TableCell>Last name</TableCell>
          <TableCell>Phone number</TableCell>
          <TableCell>Address</TableCell>
          <TableCell align="right">Actions</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {Object.entries(all).map(
          ([id, { firstName, lastName, phoneNumber, address }], index) => (
            <TableRow key={id}>
              <TableCell>{index + 1}</TableCell>
              <TableCell>{firstName}</TableCell>
              <TableCell>{lastName}</TableCell>
              <TableCell>{phoneNumber}</TableCell>
              <TableCell>{address}</TableCell>
              <TableCell align="right">
                <IconButton aria-label="edit" onClick={() => onModify(id)}>
                  <EditIcon />
                </IconButton>
                <IconButton aria-label="delete" onClick={() => onDelete(id)}>
                  <DeleteIcon />
                </IconButton>
              </TableCell>
            </TableRow>
          )
        )}
      </TableBody>
    </Table>
  );
};
