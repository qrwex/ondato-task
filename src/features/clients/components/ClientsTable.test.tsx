import { ClientsTable } from "./ClientsTable";
import { fireEvent, render } from "@testing-library/react";
import { RootState } from "../../../app/store";

describe(`${ClientsTable.name}`, () => {
  const props: any = {
    onDelete: jest.fn(),
    onModify: jest.fn(),
  };

  it("renders without crash", () => {
    const { baseElement } = render(<ClientsTable {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("handles delete click", () => {
    const { queryAllByLabelText } = render(<ClientsTable {...props} />);

    fireEvent.click(queryAllByLabelText("delete")[0]);
    expect(props.onDelete).toBeCalledWith("1");

    fireEvent.click(queryAllByLabelText("delete")[1]);
    expect(props.onDelete).toBeCalledWith("2");
  });

  it("handles modify click", () => {
    const { queryAllByLabelText } = render(<ClientsTable {...props} />);

    fireEvent.click(queryAllByLabelText("edit")[0]);
    expect(props.onModify).toBeCalledWith("1");

    fireEvent.click(queryAllByLabelText("edit")[1]);
    expect(props.onModify).toBeCalledWith("2");
  });
});

jest.mock("../../../app/hooks", () => ({
  useAppSelector: (): RootState["client"]["all"] => ({
    "1": {
      firstName: "foo",
      lastName: "bar",
      phoneNumber: "911",
      address: "mars",
    },
    "2": {
      firstName: "jr. foo",
      lastName: "bar",
      phoneNumber: "112",
      address: "5451",
    },
  }),
}));

jest.mock("@material-ui/core", () => ({
  IconButton: "icon-button",
  Table: "table",
  TableBody: "tbody",
  TableCell: "td",
  TableHead: "thead",
  TableRow: "tr",
}));

jest.mock("@material-ui/icons", () => ({
  Edit: "svg",
  Delete: "svg",
}));
