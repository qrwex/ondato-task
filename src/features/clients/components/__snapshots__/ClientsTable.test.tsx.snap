// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientsTable renders without crash 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <td>
            #
          </td>
          <td>
            First name
          </td>
          <td>
            Last name
          </td>
          <td>
            Phone number
          </td>
          <td>
            Address
          </td>
          <td
            align="right"
          >
            Actions
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            foo
          </td>
          <td>
            bar
          </td>
          <td>
            911
          </td>
          <td>
            mars
          </td>
          <td
            align="right"
          >
            <icon-button
              aria-label="edit"
            >
              <svg />
            </icon-button>
            <icon-button
              aria-label="delete"
            >
              <svg />
            </icon-button>
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            jr. foo
          </td>
          <td>
            bar
          </td>
          <td>
            112
          </td>
          <td>
            5451
          </td>
          <td
            align="right"
          >
            <icon-button
              aria-label="edit"
            >
              <svg />
            </icon-button>
            <icon-button
              aria-label="delete"
            >
              <svg />
            </icon-button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;
